src/txadmission.cpp

```diff
@@ -253,31 +243,76 @@ void CommitTxToMempool()
             SyncWithWallets(data.entry.GetSharedTx(), nullptr, -1);
 #endif
             vWhatChanged.push_back(data.hash);
-
             // Update txn per second only when a txn is valid and accepted to the mempool
             mempool.UpdateTransactionsPerSecond();
         }
         txCommitQ.clear();
     }
 
+    std::map<uint256, CTxInputData> mapWasDeferred;
     {
         LOCK(csTxInQ);
         // Clear the filter of incoming conflicts, and put all queued tx on the deferred queue since they've been
         // deferred
+        LOG(MEMPOOL, "txadmission incoming filter reset.  Current txInQ size: %d\n", txInQ.size());
         incomingConflicts.reset();
         while (!txInQ.empty())
         {
             txDeferQ.push(txInQ.front());
             txInQ.pop();
         }
-        LOG(MEMPOOL, "Reset incoming filter\n");
+
+        // Move the previously deferred txns into active processing.
+
+        // We MUST push the first item in the defer queue in without checking it against incoming conflicts.
+        // This is fine because the first insert into an empty incomingConflicts must succeed.
+        // A transaction's inputs could cause a false positive match against each other.  By pushing the first
+        // deferred tx without checking, we can still use the efficient fastfilter checkAndSet function for most queue
+        // filter checking but mop up the extremely rare tx whose inputs have false positive matches here.
+        if (!txDeferQ.empty())
+        {
+            const CTxInputData &first = txDeferQ.front();
+
+            for (auto inp : first.tx->vin)
+            {
+                uint256 hash = IncomingConflictHash(inp.prevout);
+                incomingConflicts.insert(hash);
+            }
+            txInQ.push(first);
+            cvTxInQ.notify_one();
+            txDeferQ.pop();
+        }
+
+        // Use a map to store the txns so that we end up removing duplicates which could have arrived
+        // from re-requests.
+        LOG(MEMPOOL, "popping txdeferQ, size %d\n", txDeferQ.size());
+        // this could be a lot more efficient
+        while (!txDeferQ.empty())
+        {
+            const uint256 &hash = txDeferQ.front().tx->GetHash();
+            mapWasDeferred.insert(std::pair<uint256, CTxInputData>(hash, txDeferQ.front()));
+
+            txDeferQ.pop();
+        }
+    }
+
+    if (!mapWasDeferred.empty())
+        LOG(MEMPOOL, "%d tx were deferred\n", mapWasDeferred.size());
+
+    for (auto &it : mapWasDeferred)
+    {
+        LOG(MEMPOOL, "attempt enqueue deferred %s\n", it.first.ToString());
+        EnqueueTxForAdmission(it.second);
     }
     ProcessOrphans(vWhatChanged);
 }
```

Comment:
processing orphans will now come under the cs_main lock which will increase contention for cs_main.  We don't need cs_main to process orphans so it would be better to remain at the end of ThreadCommitToMempool() where it was prior.

Reply:
I didn't move processorphans, but you are right it would be better if we get it out of cs_main